From cd75428af79ae233c6051837fa4daf43a22282be Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Thu, 20 Feb 2025 11:39:06 -0800
Subject: [PATCH 293/327] server: Use epoll_pwait2 for the main loop on Linux.

This allows higher precision waits, to match the other platforms where
we can already use a timespec directly.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=57849
---
 configure.ac | 1 +
 server/fd.c  | 8 +++++++-
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index b2b4d6612b9..3df3653b00a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2115,6 +2115,7 @@ AC_CHECK_FUNCS(\
         dladdr1 \
 	dlinfo \
 	epoll_create \
+	epoll_pwait2 \
 	fstatfs \
 	futimens \
 	futimes \
diff --git a/server/fd.c b/server/fd.c
index 1dbe8a6e20c..7e803615811 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -581,6 +581,7 @@ static inline void remove_epoll_user( struct fd *fd, int user )
 static inline void main_loop_epoll(void)
 {
     int i, ret, timeout;
+    struct timespec ts;
     struct epoll_event events[128];
 
     assert( POLLIN == EPOLLIN );
@@ -592,12 +593,17 @@ static inline void main_loop_epoll(void)
 
     while (active_users)
     {
-        timeout = get_next_timeout( NULL );
+        timeout = get_next_timeout( &ts );
 
         if (!active_users) break;  /* last user removed by a timeout */
         if (epoll_fd == -1) break;  /* an error occurred with epoll */
 
+#ifdef HAVE_EPOLL_PWAIT2
+        ret = epoll_pwait2( epoll_fd, events, ARRAY_SIZE( events ), timeout == -1 ? NULL : &ts, NULL );
+#else
         ret = epoll_wait( epoll_fd, events, ARRAY_SIZE( events ), timeout );
+#endif
+
         set_current_time();
 
         /* put the events into the pollfd array first, like poll does */
-- 
2.51.0

